### ISSUE TITLE
Cohort Clickhouse Table revamp

### ISSUE THREAD

**Issue by EDsCODE** [PostHog/posthog#13147]
**Is your feature request related to a problem?**
In an attempt to address:

- #13081 
- https://github.com/PostHog/posthog/issues/8615
- Cohort based on groups
- Related big ticket: https://github.com/PostHog/product-internal/issues/386

I've discovered some design blockers that need to be fixed.

**Issues:**
- Static cohorts and behavioral cohorts are stored in separate tables on clickhouse due to changing requirements and schemas since the start of our implementation. 
- The static cohort table does not have a version column so deletions can't happen and updates cannot happen
- The person_id field on static cohorts and behavioral cohorts are of type UUID which means it cannot store group_ids since these are generic strings
- https://posthog.slack.com/archives/C046D7C654H/p1680720872151839

**Proposal options:**
1. Create a new table `cohort_actors` and perform async migration of data from `person_static_cohort` and `cohortpeople` into this new table. The table should have the same schema as the existing `cohortpeople` table except the `person_id` field will be renamed to `actor_id` to represent persons and groups.
2. Add a new column `actor_id` that's of type `string` to existing `cohortpeople` and drop `person_id` when fully rewritten. Rename `cohortpeople` to `cohort_actors`. Backfill from `person_static_cohort` into `cohortpeople`

**Considerations:**
- The changes above will allow us to address the immediate blockers and enable better functionality across behavioral and static cohorts along with group based cohorts
- Something worth considering before commencing the above is any possible architectural changes to the schema to allow faster cohorting

WIP that led to the above:
- https://github.com/PostHog/posthog/pull/13052
- https://github.com/PostHog/posthog/pull/13104

**Comment by neilkakkar** [comment]
(I'll spend some time getting into the weeds of this tomorrow. Great issue, thanks for the context!)

Some more considerations:

1. Cohorts need to be fast at query time, but being slow at creation time is fine. We want to optimise the read-path, not the write-path.
2. Point of contention for (1) is updates. Is it a problem if cohort recalculation becomes extremely slow?
3. ???

**Comment by EDsCODE** [comment]
Good points.

1. Agreed. I'm not too sure off the top of my head what we can do to the schema here to make the read even faster
2. Indeed a problem. More frequent/realtime calculations have been asked for RE: https://github.com/PostHog/product-internal/issues/386

**Comment by neilkakkar** [comment]
One way out there idea (and not well thought through) I had for (1):

Expose cohorts a person belongs to as a materialised person property column , so we never join, nor need the `cohortpeople` table, and do update operations on the person table instead. Removes a full table join, enables fast filtering in queries as we reduce cohort subqueries to a person property filter.

No clue how this will work with person-on-events though.

**Comment by EDsCODE** [comment]
`Expose cohorts a person belongs to as a materialised person property column `
What does this column store? What if a person belongs to 100 cohorts? Is this an array of cohort ids

Another idea for better calculations, if a cohort based on persons doesn't change between recalculations, we can just only recalculate for the persons that have appeared after the last calculated timestamp.

**Comment by neilkakkar** [comment]
Correct, it's an array of cohort IDs

**Comment by EDsCODE** [comment]
Clarification as I don't think I fully caught the intent. What's the problem being solved with a new column on persons table? I don't think the storage of cohort_actor relationships is detrimental to performance at all. Let's assume that we do put cohorts in a column on the person table, every time we query for persons in a cohort, we're going to filter the persons table for persons in a cohort and then in the event query do something like ID IN (cohort_query). It doesn't matter if the filtered query is persons table with a new column or the cohort_actors table. I guess it reduces the subquery in some cases. With person on events, we always filter like e.person_id IN (cohort_query) so that subquery would exist in either scenario.

My guess is that the main problem is eventually ID IN (cohort_query) will be slow if cohort_query has millions of cohort members. Though it seems like even with some of our cohorts right now that have 1m+, there's no obvious performance loss

**Comment by macobo** [comment]
I haven't put much thought into the schema yet, but one thing that might help speed up the `IN (cohort_subquery)` subqueries are bitmaps. I haven't dug into this much (yet) but consider this slide:

![image](https://user-images.githubusercontent.com/148820/206129465-26715c0d-c0da-44bb-971d-bb78a575dd67.png)

This requires extra digging - but perhaps the underlying table can contain aggregated data somehow? If you've read the ClickHouse manual, it goes in deep with `app_metrics` table - could we do a similar aggregation trick there to store bitmaps for cohorts?

Might make it tricky to deal with people falling _out_ of cohorts though, but definetly worth an explore.

**Comment by EDsCODE** [comment]
Interesting

Example of [a query based on bitmap](https://metabase.posthog.net/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7InR5cGUiOiJuYXRpdmUiLCJuYXRpdmUiOnsicXVlcnkiOiJTRUxFQ1QgY291bnQoKikgRlJPTSBldmVudHMgd2hlcmUgdGVhbV9pZCA9IDIgYW5kIHRpbWVzdGFtcCA-ICcyMDIyLTExLTAxJyBBTkQgaGFsZk1ENShwZXJzb25faWQpIElOIChcbiAgICBTRUxFQ1QgaWRcbiAgICBGUk9NIChcbiAgICAgICAgU0VMRUNUIGJpdG1hcFRvQXJyYXkoZ3JvdXBCaXRtYXBTdGF0ZShoYXNoZWRfaWQpKSBhcyBpZHMgRlJPTSAoXG4gICAgICAgICAgICBTRUxFQ1QgaGFsZk1ENShwZXJzb25faWQpIGFzIGhhc2hlZF9pZFxuICAgICAgICAgICAgRlJPTSBjb2hvcnRwZW9wbGVcbiAgICAgICAgICAgIFdIRVJFIHRlYW1faWQgPSAyIEFORCBjb2hvcnRfaWQgPSAyMDkwNFxuICAgICAgICAgICAgR1JPVVAgQlkgcGVyc29uX2lkLCBjb2hvcnRfaWQsIHRlYW1faWQsIHZlcnNpb25cbiAgICAgICAgICAgIEhBVklORyBzdW0oc2lnbikgPiAwXG4gICAgICAgIClcbiAgICApXG4gICAgQVJSQVkgSk9JTiBpZHMgQVMgaWRcbikiLCJ0ZW1wbGF0ZS10YWdzIjp7fX0sImRhdGFiYXNlIjoyfSwiZGlzcGxheSI6InNjYWxhciIsInBhcmFtZXRlcnMiOltdLCJ2aXN1YWxpemF0aW9uX3NldHRpbmdzIjp7fX0=). 

Features:
- Need to convert uuid to uint128
- This may allow storing aggregated cohorts as suggested. So after a cohort is calculated, instead of needing to insert X number of rows. Not yet clear to me if storing bitmap objects are actually any better
- I'm not sure how to avoid the IN statement because in event queries, we can't directly do bitmapAnd as we need to also return other columns from. the event table ungrouped

**Comment by macobo** [comment]
Yeah that query style makes no difference since we're converting the bitmap back to an array - same thing, more steps.

This query works bit better:

```sql
WITH (
    SELECT groupBitmapState(id)
    FROM (
        SELECT halfMD5(person_id) AS id
        FROM cohortpeople
        WHERE team_id = 2 AND cohort_id = 20904
        GROUP BY person_id, cohort_id, team_id, version
        HAVING sum(sign) > 0
    )
) AS cohortBitmap
SELECT count(*) FROM events where team_id = 2 and timestamp > '2022-11-01' AND bitmapContains(cohortBitmap, halfMD5(person_id))
```

However it's slower than the original IN query because it needs to jump through extra hoops to build the bitmap/hash things and the resulting state is _huge_.

I can imagine an AggregatingMergeTree already storing bitmaps would be fast at this query - happy to help benchmark this if needed, but we'd run into product questions on how to _remove_ users from cohorts in nice ways. Not 100% sure that this would _actually_ be better than what IN currently does though.

In general the current approach to cohorts is not set up to be successful:
- Needing to do GROUP BY in the cohort query is fundamentally problematic as it makes _any_ query unbearably slow as cohorts get bigger - see problems with large teams and person table.
- from experience in team pipeline using collapsing merge trees is a recipe for data consistency issues
- updating and keeping it up to date is expensive and probably translates to a lot of excess load

As to how to fix these issues? :shrug: I think we should work out expected semantics from the product side and _also_ ensure it is aligned with our capabilities perf-wise.

Is this in scope for this work? :shrug: up to you, we likely can kick this problem further down the line. Tacking these notes here since cohorts are for sure also on the perf chopping block in general.

**Comment by neilkakkar** [comment]
An option for performance is to use the set or join special table engines: https://clickhouse.com/docs/en/engines/table-engines/special/set

Don't know yet which will work better, but I suspect maybe `set`.

And we populate these for cohorts that are in use, so the 2nd-nth query is fast. This doesn't help with the underlying table schema though 😅 . 

Hmm, I think the annoying bit read-wise is that `cohortpeople` is _supposed_ to be the computed people, but we still end up having to do an aggregation there to get _currently valid_ people. 

Edit: Karl mentioned this OOMed for `person_distinct_id2` table when he tried it earlier

**Comment by neilkakkar** [comment]
Another option might be to test out set / bloom filter indexes for person ID on the events table.

https://clickhouse.com/docs/en/engines/table-engines/mergetree-family/mergetree#available-types-of-indices

Should be able to aggressively discard parts where personID doesn't match the cohort personID filters.

(maybe relevant: https://github.com/PostHog/posthog/issues/12909 )

---

Hmm, another option is to leverage materialised views for the cohortpeople table. I'm not sure how much better this gets, if at all, but I can test this by having a static cohort of a million, and a dynamic cohort of a million, and checking the perf differerences between them.

**Comment by EDsCODE** [comment]
Product goals:
- There are behavioral and static cohorts
- Behavioral cohorts work as they do now along with
     - can create cohorts by groups (I want to be able to create a cohort of companies with X property)
     - have closer to real time calculation. As posthog grows, the recalculation time is only going to get longer unless we increase workers
     - be able to support the saving of arbitrary insight person results (ex: cohort of users who converted in step 2 of this funnel)
- Static cohorts
     - be able to easily add and delete 
     - upload by CSV and match based on a user defined property not just distinct_id (atleast email)
- General product requirements
     - Stored in a such a way that we can extend functionality such as perform correlation analysis on members of a cohort. (This should be possible right now)
     - Be able to see historical membership which means we'd want to consider how to include information about _when_ a user first matches the cohort conditions

> Hmm, another option is to leverage materialised views for the cohortpeople table. 

Would it be madness to sort at ingestion time? Have processing that tries all possible cohort match conditions at ingestion time and then insert with the right association as they're matched. Even when teams have 100s of cohorts, we can stack the condition matching in a way that's as efficient as possible. For example, if a cohort is AND'd across multiple conditions, only try one condition and move on to next one if this fails. Still would need to do atleast one initial heavy query and there would be duplicates. This could be a data consistency nightmare too

**Comment by EDsCODE** [comment]
> However it's slower than the original IN query because it needs to jump through extra hoops to build the bitmap/hash things and the resulting state is huge.

Sorry should've clarified that yes we would definitely want to have the bitmap already stored and not created on the fly otherwise we're just jumping through more hoops

**Comment by EDsCODE** [comment]
Given the product requirements, I think we could be differentiating implementations between static cohorts and behavioral cohorts in a better way than we are now
- Behavioral cohorts could be backed by the bitmapping aggregate tables. Deletes don't need to happen here! 
- Static cohorts can use the collapsingmergetree system that is used by behavioral cohorts right now. I think it's ok to do this because manually created cohorts will probably never be used for millions of member cohorts so inserting a few hundred to thousands of rows + managing their deletes using collapsing could be ok (as evidenced by it working fairly well for all cohorts atm).

**Comment by neilkakkar** [comment]
Did some benchmarking (results here: https://posthog.slack.com/archives/C034XD440RK/p1670935956298769 )

TL;DR: The direct query mode (no aggregation) leads to a 10x improvement in speed (from 5s -> 500ms).

It sort-of proves this hypothesis:

> Hmm, I think the annoying bit read-wise is that cohortpeople is supposed to be the computed people, but we still end up having to do an aggregation there to get currently valid people.

---

query wise, we can keep things simple with `personID IN ()` queries, and have a new schema in place which doesn't need aggregation to get the right results.

**Comment by Twixes** [comment]
@neilkakkar @EDsCODE Are there any actions we should still take based on this issue, or can we close?

**Comment by neilkakkar** [comment]
Still need this, table revamp is pending
